src/lib.rs

```diff
@@ -0,0 +1,309 @@
+        }
+
+        pub fn vote(origin, hash: T::Hash, in_favour: bool) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(Self::is_validator(&who), Error::<T>::ValidatorInvalid);
+
+            // Check if proposal exists
+            if let Some(votes) = <Votes<T>>::get(hash) {
+                // Vote if they haven't already
+                if in_favour {
+                    Self::vote_for(who, votes)?
+                } else {
+                    Self::vote_against(who, votes)?
+                }
+            } else {
+                Err(Error::<T>::ProposalDoesNotExist)?
+            }
+
+            Ok(())
+        }
+
+        /// Completes an asset transfer to the chain by emitting an event to be acted on by the
+        /// bridge and updating the tx count for the respective chan.
+        pub fn receive_asset(origin, dest_id: Vec<u8>, to: Vec<u8>, token_id: Vec<u8>, metadata: Vec<u8>) -> DispatchResult {
+            // TODO: Limit access
+            ensure_root(origin)?;
+            // Ensure chain is whitelisted
+            if let Some(mut counter) = Chains::get(&dest_id) {
+                // Increment counter and store
+                counter.recv += 1;
+                Chains::insert(&dest_id, counter.clone());
+                Self::deposit_event(RawEvent::AssetTransfer(dest_id, counter.recv, to, token_id, metadata));
+                Ok(())
+            } else {
+                Err(Error::<T>::ChainNotWhitelisted)?
+            }
+        }
+
+        // TODO: Should use correct amount type
+        pub fn transfer(origin, to: T::AccountId, amount: u32) -> DispatchResult {
+            ensure_root(origin)?;
+            let source: T::AccountId = <EndowedAccount<T>>::get();
+            T::Currency::transfer(&source, &to, amount.into(), AllowDeath)?;
+            Ok(())
+        }
+    }
+}
+
+/// Main module declaration.
+/// Here we should include non-state changing public funcs
+impl<T: Trait> Module<T> {
+    pub fn is_validator(who: &T::AccountId) -> bool {
+        Self::validators(who)
+    }
+
+    fn initialize_validators(validators: &[T::AccountId]) {
+        if !validators.is_empty() {
+            for v in validators {
+                <Validators<T>>::insert(v, true);
+            }
+        }
+    }
+
+    /// Note: Existence of proposal must be checked before calling
+    fn vote_for(
+        who: T::AccountId,
+        mut votes: ProposalVotes<T::AccountId, T::Hash>,
+    ) -> DispatchResult {
+        //let mut prop = <Proposals<T>>::get((deposit_id, call.clone())).unwrap();
+        if !votes.votes_for.contains(&who) {
+            votes.votes_for.push(who.clone());
+            <Votes<T>>::insert(votes.hash, votes.clone());
+            Self::deposit_event(RawEvent::VoteFor(votes.hash, who.clone()));
+
+            if votes.votes_for.len() == <ValidatorThreshold>::get() as usize {
+                Self::finalize_transfer(votes)?
```

Comment:
Removethe proposal here? Keep in mind that Substrate does not role back any changes to storage – so make sure to ensure that the proposal exists before changing state.